SUB3/app.py

```diff
@@ -175,36 +188,64 @@ def show_m1display(position):
+            logger.error("Illegal state: not frame.paused and not frame.running. Corrected to frame.paused and not frame.running.")
+            frame.paused = True
         if frame.paused:
+            if frame.running:
+                # Running, but paused
+                other_opts_btn['state'] = 'normal'
+                start_btn['state'] = 'disabled'
+                pause_btn['state'] = 'normal'
+                stop_btn['state'] = 'normal'
+
+                if pause_btn_color_swap and time.time() - swap_time > PAUSE_BLINK_RATE:
+                    pause_btn_color_swap = not pause_btn_color_swap
+                    pause_btn.configure(bg="red")
+                    swap_time = time.time()
+
+                elif not pause_btn_color_swap and time.time() - swap_time > PAUSE_BLINK_RATE:
+                    pause_btn_color_swap = not pause_btn_color_swap
+                    pause_btn.configure(bg="green")
+                    swap_time = time.time()
+            else:
+                # Not running; stopped
+                other_opts_btn['state'] = 'normal'
+                start_btn['state'] = 'normal'
+                pause_btn['state'] = 'disabled'
+                stop_btn['state'] = 'disabled'
 
-            other_opts_btn['state'] = 'normal'
-
-            if pause_btn_color_swap and time.time() - swap_time > PAUSE_BLINK_RATE:
-                pause_btn_color_swap = not pause_btn_color_swap
                 pause_btn.configure(bg="red")
-                swap_time = time.time()
-
-            elif not pause_btn_color_swap and time.time() - swap_time > PAUSE_BLINK_RATE:
-                pause_btn_color_swap = not pause_btn_color_swap
-                pause_btn.configure(bg="green")
-                swap_time = time.time()
         else:
-            if not frame.running:
-                other_opts_btn['state'] = 'normal'
-            else:
-                other_opts_btn['state'] = 'disabled'
+            # Running, not paused
+            other_opts_btn['state'] = 'disabled'
+            start_btn['state'] = 'disabled'
+            pause_btn['state'] = 'normal'
+            stop_btn['state'] = 'normal'
 
             pause_btn.configure(bg="red")
 
         # Check for updates and then change values
-        if options.updates:
-            #Update preload values
-            frame.update_preloads(options.pre_min,options.pre_max)
-            preload_display.update_preloads(options.pre_min,options.pre_max)
+        if options["updates"]:
+            # Update preload values
+            frame.update_preloads(options["pre_min"], options["pre_max"])
+            preload_display.update_preloads(options["pre_min"], options["pre_max"])
+
+            # Update M1 bounds
+            m1_display.update_all(m1min=options["m1_min"], m1max=options["m1_max"])
+
+            # Update block
+            if options["pat_id"] != frame.block.patID:
```

Comment:
Just a thought, we might consider moving these updates to frame and letting frame manage it's own information. This could be paired with the frame.update_preloads into a generic function for all updates. I like how you did update_all for m1_display. It keeps the boundaries clear between the different classes.